### PR TITLE
config: Add config test for C11 _Static_assert.

### PR DESCRIPTION
**Pull request by hzhou**
This is a fix for issue #3178.

Removed adhoc tests in mpir_assert.h. 
Added PAC_C_STATIC_ASSERT in aclocal_cc.m4. 
Updated configure.ac to run that test.

### CODE REVIEW

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
There is also CH4_COMPILE_TIME_ASSERT defined in src/mpid/ch4/include/mpidpre.h, which may also can use this _Static_assert test.

File: src/mpl/include/mpl_base.h [hunk 1]
```diff
@@ -98,6 +98,18 @@
 #define likely(x_)   (x_)
 #endif
 
+#ifdef HAVE_C11__STATIC_ASSERT
+#define MPL_Static_assert(cond_,msg_) _Static_assert(cond_,msg_)
```
Comment (raffenet) [inline src/mpl/include/mpl_base.h]: New additions to MPL have used all lowercase after `MPL_`. I think that would be preferred here.

File: confdb/aclocal_cc.m4 [hunk 2]
```diff
@@ -1696,3 +1696,23 @@ if test x$have_builtin_expect = xyes ; then
     AC_DEFINE([HAVE_BUILTIN_EXPECT], [1], [Define to 1 if the compiler supports __builtin_expect.])
 fi
 ])
+
+dnl
+dnl PAC_C_STATIC_ASSERT - Test whether C11 _Stattic_assert is supported
```
Comment (raffenet) [inline confdb/aclocal_cc.m4]: Typo Stattic->Static

File: src/mpl/include/mpl_base.h [hunk 3]
```diff
@@ -98,6 +98,18 @@
 #define likely(x_)   (x_)
 #endif
 
+#ifdef HAVE_C11__STATIC_ASSERT
+#define MPL_static_assert(cond_,msg_) _Static_assert(cond_,msg_)
```
Comment (raffenet) [inline src/mpl/include/mpl_base.h]: Make sure the commit message reflects the case change we made in the code. `MPL_Static_assert` -> `MPL_static_assert`.

File: src/mpl/include/mpl_base.h [hunk 4]
```diff
@@ -98,6 +98,18 @@
 #define likely(x_)   (x_)
 #endif
 
+#ifdef HAVE_C11__STATIC_ASSERT
```
Comment (raffenet) [inline src/mpl/include/mpl_base.h]: MPL's config output is prefixed with `MPL_`, so I think this check needs to be modified.

> /* Define if C11 _Stattic_assert is supported. */
> #ifndef MPL_HAVE_C11__STATIC_ASSERT
> #define MPL_HAVE_C11__STATIC_ASSERT  1
> #endif

**Comment by raffenet** [comment]
@hzhou looks like this needs rebasing. Not sure how the merge commit ended up in here, but it should be removed, as well.

**Comment by hzhou** [comment]
@raffenet There was a merge conflict (src/mpid/ch4/netmod/ofi/ofi_win.h) ... probably because I got new computer and merged PR locally.  Reverted.

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Comment by raffenet** [comment]
Will merge after tests complete successfully.

**Comment by raffenet** [comment]
Looks like this doesn't actually compile because each instance of `MPL_static_assert` is missing the message argument. I don't know if we need to actually go around adding messages. Maybe just create a default?

**Comment by hzhou** [comment]
I created another macro `MPL_COMPILE_TIME_ASSERT` that accepts a single argument. It just uses a default error message (MPL_COMPILE_TIME_ASSERT failed).

**Comment by hzhou** [comment]
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch4/most

File: src/mpl/include/mpl_base.h [hunk 5]
```diff
@@ -116,6 +116,8 @@
     do { switch(0) { case 0: case (expr_): default: break; } } while (0)
 #endif
 
+#define MPL_COMPILE_TIME_ASSERT(cond_) MPL_static_assert(cond_, "MPL_COMPILE_TIME_ASSERT failure")
```
Comment (raffenet) [inline src/mpl/include/mpl_base.h]: I think the addition of this new macro in MPL should be part of the previous MPL commit.

Comment (hzhou) [reply src/mpl/include/mpl_base.h]: The single `cond_` assert are added only to replace `{CH4,OFI}_COMPILE_TIME_ASSERT`, therefore I found it more relevant to group them together in a single commit. But I do not insist.

Comment (hzhou) [reply src/mpl/include/mpl_base.h]: Fair enough. I'll update it later.

**Comment by hzhou** [comment]
test:jenkins/ch4/most

**Pull request merged**